File: baekjoon/python/gold4/q8983_hunter.py
## https://www.acmicpc.net/problem/8983

import sys
input = sys.stdin.readline

def sol_fail():
    m, n, l = map(int, input().split())
    shooting_points = list(map(int, input().split()))
    shooting_points.sort()
    animal_points = []
    for _ in range(n):
        x, y = map(int, input().split())
        if y <= l and (shooting_points[0] - l <= x <= shooting_points[-1]+l):
            animal_points.append((x, y))

    animal_points.sort()
    ans = 0
    idx = 0
    for animal in animal_points:
        start, end = idx, m-1
        while start <= end:
            mid = (start + end) // 2
            if shooting_points[mid] < animal[0]:
                start = mid + 1
            else:
                end = mid - 1
                idx = mid

        if abs(shooting_points[idx] - animal[0]) + animal[1] <= l :
            ans += 1
        elif idx != 0 and abs(shooting_points[idx-1] - animal[0]) + animal[1] <= l :
            ans += 1

    print(ans)


def sol():
    m, n, l = map(int, input().split())
    shooting_points = list(map(int, input().split()))
    shooting_points.sort()
    animal_points = []
    for _ in range(n):
        x, y  = map(int, input().split())
        animal_points.append((x, y))

    ans = 0
    for animal in animal_points:
        if animal[1] > l:
            continue

        start, end = 0, m-1
        point = m
        diff = l + 1

        while start <= end:
            mid = (start + end) // 2
            d = shooting_points[mid] - animal[0]

            if abs(d) < diff:
                diff = abs(d)
                point = mid

            if d < 0:
                start = mid + 1
            else:
                end = mid - 1
        
        if(abs(shooting_points[point] - animal[0]) + animal[1]) <= l:
            ans +=1 
        print(animal, point, ans)

    print(ans)


sol()

File: baekjoon/java/src/gold4/Q8983_Hunter.java
package gold4;


import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Q8983_Hunter  {
    static class Position {
        int x, y;

        public Position(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static void input() {
        M = scan.nextInt();
        N = scan.nextInt();
        L = scan.nextInt();
        shots = new int[M];
        for (int i = 0; i < M; i++) {
            shots[i] = scan.nextInt();
        }
        animals = new Position[N];
        for (int i = 0; i < N; i++) {
            int x = scan.nextInt();
            int y = scan.nextInt();
            animals[i] = new Position(x, y);
        }
    }

    static int M, N, L;
    static int[] shots;
    static Position[] animals;

    static void sol() {
        int ans = 0;
        Arrays.sort(shots);
        for (Position animal : animals) {
            if(animal.y > L) {
                continue;
            }
            int start = 0;
            int end = shots.length -1;
            int minLength = L + 1;
            int pos = -1;
            while(start <= end) {
                int mid = (start + end) / 2;
                int dis = shots[mid] - animal.x;

                if(Math.abs(dis) < minLength) {
                    minLength = Math.abs(dis);
                    pos = mid;
                }

                if(dis < 0) {
                    start = mid + 1;
                } else {
                    end = mid - 1;
                }
            }

            if(pos != -1 && Math.abs(shots[pos]-animal.x) + animal.y <= L) {
                ans += 1;
            }
        }
        System.out.println(ans);
    }



    public static void main(String[] args) {
        input();
        sol();
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}
